src/lex.cpp

```diff
@@ -1288,6 +1292,52 @@ void lexer::parse_modern_octal_number() {
       diag_unexpected_characters_in_octal_number>(input);
 }
 
+void lexer::check_precision_loss(const char8* number_begin,
+                                 const char8* input) {
+  string8_view number_literal(number_begin,
+                              static_cast<size_t>(input - number_begin));
```

Comment:
Nit: Prefer `narrow_cast` over `static_cast` when narrowing. `narrow_cast` will crash (instead of wrap) if the input is out of bounds.
```suggestion
                              narrow_cast<size_t>(input - number_begin));
```